linalg/dual.hpp

```diff
@@ -0,0 +1,413 @@
+dual<value_type, gradient_type> asin(dual<value_type, gradient_type> a)
+{
+   using std::sqrt;
+   using std::asin;
+   return {asin(a.value), a.gradient / sqrt(value_type{1} - a.value * a.value)};
+}
+
+/** @brief implementation of tan for dual numbers */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> tan(dual<value_type, gradient_type> a)
+{
+   using std::tan;
+   value_type f = tan(a.value);
+   return {f, a.gradient * (value_type{1} + f * f)};
+}
+
+/** @brief implementation of atan for dual numbers */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> atan(dual<value_type, gradient_type> a)
+{
+   return {atan(a.value), a.gradient / (value_type{1} + a.value * a.value)};
+}
+
+/** @brief implementation of exponential function for dual numbers */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> exp(dual<value_type, gradient_type> a)
+{
+   return {std::exp(a.value), std::exp(a.value) * a.gradient};
+}
+
+/** @brief implementation of the natural logarithm function for dual numbers */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> log(dual<value_type, gradient_type> a)
+{
+   return {std::log(a.value), a.gradient / a.value};
+}
+
+/** @brief implementation of `a` (dual) raised to the `b` (dual) power */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> pow(dual<value_type, gradient_type> a,
+                                    dual<value_type, gradient_type> b)
+{
+   value_type value = pow(a.value, b.value);
+   return {value, value * (a.gradient * (b.value / a.value) + b.gradient * std::log(a.value))};
+}
+
+/** @brief implementation of `a` (non-dual) raised to the `b` (dual) power */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> pow(double a, dual<value_type, gradient_type> b)
+{
+   value_type value = pow(a, b.value);
+   return {value, value * b.gradient * std::log(a)};
+}
+
+/** @brief implementation of `a` (non-dual) raised to the `b` (non-dual) power */
+template <typename value_type > MFEM_HOST_DEVICE
+value_type pow(value_type a, value_type b) { return std::pow(a, b); }
+
+/** @brief implementation of `a` (dual) raised to the `b` (non-dual) power */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> pow(dual<value_type, gradient_type> a, double b)
+{
+   value_type value = pow(a.value, b);
+   return {value, value * a.gradient * b / a.value};
+}
+
+/** @brief overload of operator<< for `dual` to work with work with standard output streams */
+template <typename value_type, typename gradient_type, int... n>
+std::ostream& operator<<(std::ostream& out, dual<value_type, gradient_type> A)
```

Comment:
Won't this shadow `mfem::out`? In any case, probably better to use `os` instead of `out` for consistency